### PR TITLE
Podman has been integrated to be used as an optional replacement for docker

### PR DESCRIPTION
**Pull request by rishabhsri18**
Changes which will pick the container engine name from the environment and replaced all the docker argument with the variable of containerEngineType

### CODE REVIEW

File: common/common.go [hunk 1]
```diff
@@ -634,12 +665,16 @@ func ExtractConfig(configs []string) map[string]string {
 
 // DockerPush pushes to docker registry.
 func DockerPush(ff *FuncFile) error {
```
Comment (theatul) [inline common/common.go:667]: we need to modify these function names.

File: common/common.go [hunk 2]
```diff
@@ -274,8 +273,24 @@ func PrintContextualInfo() {
 	fmt.Println("Current Context: ", currentContext)
 }
 
+func GetContainerEngineType() (string, error) {
```
[the remark this answers is not in the document]
Comment (rishabhsri18) [reply common/common.go:276]: We set docker as default if no info is present it's just the reference where we get the config details from viper. I will return docker here as well if we can't find any value

File: common/common.go [hunk 3]
```diff
@@ -442,17 +470,20 @@ func RunBuild(verbose bool, dir, imageName, dockerfile string, buildArgs []strin
 	return nil
 }
 
-func dockerVersionCheck() error {
-	out, err := exec.Command("docker", "version", "--format", "{{.Server.Version}}").Output()
+func dockerVersionCheck(containerEngineType string) error {
```
[the remark this answers is not in the document]
Comment (rishabhsri18) [reply common/common.go:473]: It check the container engine w.r.t the version call, we have a minimum version requirement for docker and not for podman. So just checked the podman version and docker condition remains the same

File: config/config.go [hunk 4]
```diff
@@ -247,6 +259,14 @@ func LoadConfiguration(c *cli.Context) error {
 	return nil
 }
 
+func ValidateContainerEngineType(containerEngineType string) error {
+	if strings.Compare(containerEngineType, "docker") != 0 && strings.Compare(containerEngineType, "podman") != 0 &&
+		strings.Compare(containerEngineType, "") != 0 {
+		return errors.New("Invalid Container Engine")
+	}
+	return nil
```
Comment (metamemelord) [inline config/config.go:267]: I think a switch with fallthrough will suffice.

File: common/common.go [hunk 5]
```diff
@@ -632,34 +664,43 @@ func ExtractConfig(configs []string) map[string]string {
...
+		return fmt.Errorf("error running docker/podman push, are you logged into docker?: %v", err)
 	}
 	return nil
 }
 
-// DockerPush pushes to docker registry.
-func DockerPushV20180708(ff *FuncFileV20180708) error {
-	err := ValidateFullImageName(ff.ImageNameV20180708())
+// ContainerEnginePush pushes to docker registry.
+func ContainerEnginePushV20180708(ff *FuncFileV20180708) error {
```
Comment (metamemelord) [inline common/common.go:688]: "image registry".

Can be renamed `PushImageV20180708` to be explicit.

File: common/common.go [hunk 6]
```diff
@@ -632,34 +664,43 @@ func ExtractConfig(configs []string) map[string]string {
 	return c
 }
 
-// DockerPush pushes to docker registry.
-func DockerPush(ff *FuncFile) error {
-	err := ValidateFullImageName(ff.ImageName())
+// ContainerEnginePush pushes to docker registry.
+func ContainerEnginePush(ff *FuncFile) error {
```
Comment (metamemelord) [inline common/common.go:668]: Same as last.

File: common/common.go [hunk 7]
```diff
@@ -319,8 +336,14 @@ func dockerBuild(verbose bool, fpath string, ff *FuncFile, buildArgs []string, n
...
 
-func dockerBuildV20180708(verbose bool, fpath string, ff *FuncFileV20180708, buildArgs []string, noCache bool) error {
-	err := dockerVersionCheck()
+func containerEngineBuildV20180708(verbose bool, fpath string, ff *FuncFileV20180708, buildArgs []string, noCache bool) error {
+	containerEngineType, err := GetContainerEngineType()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Using Container engine ", containerEngineType)
```
Comment (metamemelord) [inline common/common.go:345]: More like an OCD 😄 commas between `Println` add spaced automatically. I hope all these will be cleaned up before merging.

File: common/common.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (metamemelord) [reply common/common.go:276]: We can use `viper.SetDefault` to ensure `""` is never returned.

[the remark this answers is not in the document]
Comment (rishabhsri18) [reply common/common.go:276]: So the viper.SetDefault is there which is setting up the default value for the containerEngine value to docker as default

File: common/common.go [hunk 8]
```diff
@@ -632,34 +664,43 @@ func ExtractConfig(configs []string) map[string]string {
...
 		return err
 	}
 	fmt.Printf("Pushing %v to docker registry...", ff.ImageName())
-	cmd := exec.Command("docker", "push", ff.ImageName())
+	cmd := exec.Command(containerEngineType, "push", ff.ImageName())
 	cmd.Stderr = os.Stderr
 	cmd.Stdout = os.Stdout
 	if err := cmd.Run(); err != nil {
-		return fmt.Errorf("error running docker push, are you logged into docker?: %v", err)
+		return fmt.Errorf("error running docker/podman push, are you logged into docker?: %v", err)
```
Comment (roneetshaw) [inline common/common.go:682]: Can we pass the containerEngineType here?
Remove docker from the statement

File: common/common.go [hunk 9]
```diff
@@ -632,34 +664,43 @@ func ExtractConfig(configs []string) map[string]string {
...
 	}
+	fmt.Println("Using Container engine ", containerEngineType, "to push")
 	fmt.Printf("Pushing %v to docker registry...", ff.ImageNameV20180708())
-	cmd := exec.Command("docker", "push", ff.ImageNameV20180708())
+	cmd := exec.Command(containerEngineType, "push", ff.ImageNameV20180708())
 	cmd.Stderr = os.Stderr
 	cmd.Stdout = os.Stdout
 	if err := cmd.Run(); err != nil {
-		return fmt.Errorf("error running docker push, are you logged into docker?: %v", err)
+		return fmt.Errorf("error running docker/podman push, are you logged into docker?: %v", err)
```
Comment (roneetshaw) [inline common/common.go:703]: Can we pass the containerEngineType here?
Remove docker from the statement

File: common/common.go [hunk 10]
```diff
@@ -886,15 +927,19 @@ func ListTriggersInFunc(c *cli.Context, client *fnclient.Fn, fn *modelsv2.Fn) ([
...
+	}
 	fmt.Println("Running init-image: " + initImage)
 
 	args := []string{"run", "--rm", "-e", "FN_FUNCTION_NAME=" + fName}
 	args = append(args, proxyArgs()...)
 	args = append(args, initImage)
 
-	fmt.Printf("Executing docker command: %s\n", strings.Join(args, " "))
-	cmd := exec.Command("docker", args...)
+	fmt.Printf("Executing docker/podman command: %s\n", strings.Join(args, " "))
```
Comment (roneetshaw) [inline common/common.go:941]: Pass the ```containerEngineType``` in the info message

File: common/common.go [hunk 11]
```diff
@@ -886,15 +927,19 @@ func ListTriggersInFunc(c *cli.Context, client *fnclient.Fn, fn *modelsv2.Fn) ([
 	return resTriggers, nil
 }
 
-func DockerRunInitImage(initImage string, fName string) error {
+func RunInitImage(initImage string, fName string) error {
+	containerEngineType, err := GetContainerEngineType()
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (rishabhsri18) [reply common/common.go:929]: So this is basically required as we don't have any repetitive code it is different for every commands and i think for getting the data in every command we need to call this function

File: common/common.go [hunk 12]
```diff
@@ -442,17 +471,20 @@ func RunBuild(verbose bool, dir, imageName, dockerfile string, buildArgs []strin
...
 	}
 	// dev / test builds append '-ce', trim this
 	trimmed := strings.TrimRightFunc(string(out), func(r rune) bool { return r != '.' && !unicode.IsDigit(r) })
 
 	v, err := semver.NewVersion(trimmed)
 	if err != nil {
-		return fmt.Errorf("could not check Docker version: %v", err)
+		return fmt.Errorf("could not check Docker/Podman version: %v", err)
+	}
+	if containerEngineType == "podman" {
```
[the remark this answers is not in the document]
Comment (rishabhsri18) [reply common/common.go]: We can definitely use some factory pattern design to get rid of multiple checking but we need to refactor a lot of code for this

File: common/common.go [hunk 13]
```diff
@@ -141,7 +141,7 @@ func BuildFuncV20180708(verbose bool, fpath string, funcfile *FuncFileV20180708,
 		return nil, err
 	}
 
-	if err := dockerBuildV20180708(verbose, fpath, funcfile, buildArg, noCache); err != nil {
+	if err := containerEngineBuildV20180708(verbose, fpath, funcfile, buildArg, noCache); err != nil {
```
Comment (theatul) [inline common/common.go:144]: We can probably rename this ti just Build<version> or a more appropriate name in a future change.

**Review by metamemelord — APPROVED** [review]
LGTM.

**Review by roneetshaw — APPROVED** [review]

**Pull request merged**